File: ADS PROGRAMS/avl.c
#include<stdio.h>
#include<stdlib.h>
typedef struct node avl;
struct node
{
struct node *lchild,*rchild;
int data;
};
void inorder(avl *temp)
{
if(temp!=NULL)
{
inorder(temp->lchild);
printf("%d\t",temp->data);
inorder(temp->rchild);
}
}
avl* LL(avl *tree)
{
avl *k2=tree,*k1=k2->lchild;
k2->lchild=k1->rchild;
k1->rchild=k2;
return k1;
}
avl* RR(avl *tree)
{
avl *k1=tree,*k2=k1->rchild;
k1->rchild=k2->lchild;
k2->lchild=k1;
return k1;
}
avl* LR(avl *tree)
{
avl *k3=tree,*k1=k3->lchild,*k2=k1->rchild;
k1->rchild=k2->lchild;
k3->lchild=k2->rchild;
k2->lchild=k1;
k2->rchild=k3;
return k2;
}
avl *RL(avl *tree)
{
avl *k1=tree,*k3=k1->rchild,*k2=k2->lchild;
k1->rchild=k2->lchild;
k3->lchild=k2->rchild;
k2->lchild=k1;
k2->rchild=k2;
return k2;
}
avl* balance(avl* tree)
{
int heightdiff=height(tree->lchild)-height(tree->rchild);
if(heightdiff==2)
{
	int x=height(tree->lchild->lchild)-height(tree->lchild->rchild);
	if(x==1||x==0)
		tree=LL(tree);
	else
		tree=LR(tree);
}
else if(heightdiff==-2)
{
	int y=height(tree->rchild->lchild)-height(tree->rchild->rchild);
	if(y==-1||y==0)
		tree=RR(tree);
	else
		tree=RL(tree);
}
return tree;
} 
avl* insert(avl *tree,int ele)
{
if(tree==NULL)
{
	tree=(avl*)malloc(sizeof(avl));
	tree->data=ele;
	tree->rchild=tree->lchild=NULL;
}
else
{
	if(ele<tree->data)
	{
		tree->lchild=insert(tree->lchild,ele);
		tree=balance(tree);
	}
	else if(ele>tree->data)
	{
	tree->rchild=insert(tree->rchild,ele);
	tree=balance(tree);
	}
}
return tree;
}
int height(avl *tree)
{
if(tree==NULL)
return -1;
else
{
int lh=height(tree->lchild);
int rh=height(tree->rchild);
if(lh>rh)
	return lh+1;
else
	return rh+1;
}
}
int main()
{
avl *root=NULL;
int ele,i=1;
while(i<3 &&  i>0)
{
printf("\nEnter 1 to insert\nEnter 2 to display.");
scanf("%d",&i);
switch(i)
{
case 1:printf("\nEnter the element  :    ");
	    scanf("%d",&ele);
	    root=insert(root,ele);
	    break;
case 2:inorder(root);
	     break;
default:
           printf("Wrong choice");	
}
}
return 0;
}

File: PYTHON DS/avl.py
class AVL:
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None
    @staticmethod
    def insert(root,ele):
        if root is None:
            root=AVL(ele)
        elif(ele < root.data):
            root.left=root.insert(root.left,ele)
            root=root.balance(root)
        elif(ele > root.data):
            root.right=root.insert(root.right,ele)
            root=root.balance(root)
        else:
            print("Element Already Exits")
        return root
    @staticmethod
    def balance(root):
        height_diff=root.height(root.left)-root.height(root.right)
        if height_diff==2:
            x=root.height(root.left.left)-root.height(root.left.right)
            if x==1 or x==0:
                root=root.leftleft(root)
            else:
                root=root.leftright(root)
        elif height_diff==-2:
            y=root.height(root.right.left)-root.height(root.right.right)
            if y==-1 or y==0:
                root=root.rightright(root)
            else:
                root=root.rightleft(root)
        return root
    @staticmethod
    def inorder(root):
        if root is not None :
            root.inorder(root.left)
            print(root.data,end=" ")
            root.inorder(root.right)
    @staticmethod
    def preorder(root):
        if root is not None :
            print(root.data,end=" ")
            root.preorder(root.left)
            root.preorder(root.right)
    @staticmethod
    def postorder(root):
        if root is not None :
            root.postorder(root.left)
            root.postorder(root.right)
            print(root.data,end=" ")
    @staticmethod
    def height(root):
        if root is None :
            return -1
        else:
            left_height=root.height(root.left)
            right_height=root.height(root.right)
            if left_height > right_height :
                return left_height + 1
            else:
                return right_height + 1
    @staticmethod
    def leftleft(root):
        k2=root
        k1=k2.left
        k2.left=k1.right
        k1.right=k2
        return k1
    @staticmethod
    def rightright(root):
        k1=root
        k2=k1.right
        k1.right=k2.left
        k2.left=k1
        return k2
    @staticmethod
    def leftright(root):
        k3=root
        k1=k3.left
        k2=k1.right
        k1.right=k2.left
        k3.left=k2.right
        k2.left=k1
        k2.right=k3
        return k2
    @staticmethod
    def rightleft(root):
        k1=root
        k3=k1.right
        k2=k3.left
        k1.right=k2.left
        k3.left=k2.right
        k2.left=k1
        k2.right=k3
        return k2
root=None
N=int(input("Enter the Number of Values:"))
values=str(input("Enter the Values:"))
for i in values.split():
    root=AVL.insert(root,i)
print("  \nDisplay Final AVL Tree Elements in Inorder  ")
AVL.inorder(root)
print("  \nDisplay Final AVL Tree Elements in Preorder  ")
AVL.preorder(root)
print("  \nDisplay Final AVL Tree Elements in Postorder  ")
AVL.postorder(root)

    


